src/uu/ls/src/ls.rs

```diff
@@ -3195,17 +3185,58 @@ impl StyleManager {
     }
 }
 
-/// Colors the provided name based on the style determined for the given path.
+fn apply_style_based_on_metadata(
+    path: &PathData,
+    md_option: Option<&Metadata>,
+    ls_colors: &LsColors,
+    style_manager: &mut StyleManager,
+    name: &str,
+) -> String {
+    match ls_colors.style_for_path_with_metadata(&path.p_buf, md_option) {
+        Some(style) => style_manager.apply_style(style, name),
+        None => name.to_owned(),
+    }
+}
+
+/// Colors the provided name based on the style determined for the given path
+/// This function is quite long because it tries to leverage DirEntry to avoid
+/// unnecessary calls to stat()
+/// and manages the symlink errors
 fn color_name(
     name: String,
-    path: &Path,
-    md: Option<&Metadata>,
+    path: &PathData,
     ls_colors: &LsColors,
     style_manager: &mut StyleManager,
+    out: &mut BufWriter<Stdout>,
+    check_for_deref: bool,
+    target_symlink: Option<&PathData>,
 ) -> String {
-    match ls_colors.style_for_path_with_metadata(path, md) {
-        Some(style) => style_manager.apply_style(style, &name),
-        None => name,
+    if !path.must_dereference {
+        // If we need to dereference (follow) a symlink, we will need to get the metadata
+        if let Some(de) = &path.de {
+            // There is a DirEntry, we don't need to get the metadata for the color
+            return match ls_colors.style_for(de) {
+                Some(style) => style_manager.apply_style(style, &name),
+                None => name,
+            };
+        }
+    }
+
+    if check_for_deref {
+        // use the optional target_symlink
+        // Use fn get_metadata instead of md() here and above because ls
```

Comment:
The `md` function no longer exists:
```suggestion
        // Use fn get_metadata_with_deref_opt instead of get_metadata() here because ls
```